File: trunk/src/main/java/net/esper/pattern/observer/TimerAtObserver.java
package net.esper.pattern.observer;

import net.esper.pattern.PatternContext;
import net.esper.pattern.MatchedEventMap;
import net.esper.schedule.ScheduleCallback;
import net.esper.schedule.ScheduleSpec;
import net.esper.pattern.observer.EventObserver;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

/**
 * Observer implementation for indicating that a certain time arrived, similar to "crontab".
 */
public class TimerAtObserver implements EventObserver, ScheduleCallback
{
    private final ScheduleSpec scheduleSpec;
    private final PatternContext context;
    private final MatchedEventMap beginState;
    private final ObserverEventEvaluator observerEventEvaluator;

    private boolean isTimerActive = false;

    /**
     * Ctor.
     * @param scheduleSpec - specification containing the crontab schedule
     * @param context - timer serive to use
     * @param beginState - start state
     * @param observerEventEvaluator - receiver for events
     */
    public TimerAtObserver(ScheduleSpec scheduleSpec, PatternContext context, MatchedEventMap beginState, ObserverEventEvaluator observerEventEvaluator)
    {
        this.scheduleSpec = scheduleSpec;
        this.context = context;
        this.beginState = beginState;
        this.observerEventEvaluator = observerEventEvaluator;
    }

    public final void scheduledTrigger()
    {
        if (log.isDebugEnabled())
        {
            log.debug(".scheduledTrigger");
        }

        observerEventEvaluator.observerEvaluateTrue(beginState);
        isTimerActive = false;
    }

    public void startObserve()
    {
        if (log.isDebugEnabled())
        {
            log.debug(".startObserve Starting at, spec=" + scheduleSpec);
        }

        if (isTimerActive == true)
        {
            throw new IllegalStateException("Timer already active");
        }

        context.getSchedulingService().add(scheduleSpec, this);
        isTimerActive = true;
    }

    public void stopObserve()
    {
        if (log.isDebugEnabled())
        {
            log.debug(".stopObserve");
        }

        if (isTimerActive)
        {
            context.getSchedulingService().remove(this);
            isTimerActive = false;
        }
    }

    private static final Log log = LogFactory.getLog(TimerAtObserver.class);
}

File: NEsper/pattern/observer/TimerAtObserver.cs
using net.esper.core;
using net.esper.compat;
using net.esper.pattern;
using net.esper.schedule;

using org.apache.commons.logging;

namespace net.esper.pattern.observer
{
    /// <summary>
    /// Observer implementation for indicating that a certain time arrived, similar to "crontab".
    /// </summary>

    public class TimerAtObserver 
		: EventObserver
		, ScheduleHandleCallback
    {
        private readonly ScheduleSpec scheduleSpec;
        private readonly PatternContext context;
        private readonly ScheduleSlot scheduleSlot;
        private readonly MatchedEventMap beginState;
        private readonly ObserverEventEvaluator observerEventEvaluator;

        private bool isTimerActive = false;
		private EPStatementHandleCallback scheduleHandle;

        /// <summary> Ctor.</summary>
        /// <param name="scheduleSpec">specification containing the crontab schedule
        /// </param>
        /// <param name="context">timer serive to use
        /// </param>
        /// <param name="beginState">Start state
        /// </param>
        /// <param name="observerEventEvaluator">receiver for events
        /// </param>
        public TimerAtObserver(ScheduleSpec scheduleSpec, PatternContext context, MatchedEventMap beginState, ObserverEventEvaluator observerEventEvaluator)
        {
            this.scheduleSpec = scheduleSpec;
            this.context = context;
            this.beginState = beginState;
            this.observerEventEvaluator = observerEventEvaluator;
            this.scheduleSlot = context.ScheduleBucket.AllocateSlot();
        }

        /// <summary>
        /// Called when a scheduled callback occurs.
        /// </summary>
        public void ScheduledTrigger(ExtensionServicesContext extensionServicesContext)
        {
            if (log.IsDebugEnabled)
            {
                log.Debug(".scheduledTrigger");
            }

            observerEventEvaluator.ObserverEvaluateTrue(beginState);
            isTimerActive = false;
        }

        /// <summary>
        /// Start observing.
        /// </summary>
        public virtual void StartObserve()
        {
            if (log.IsDebugEnabled)
            {
                log.Debug(".StartObserve Starting at, spec=" + scheduleSpec);
            }

            if (isTimerActive == true)
            {
                throw new IllegalStateException("Timer already active");
            }

	        scheduleHandle = new EPStatementHandleCallback(context.EpStatementHandle, this);
	        context.SchedulingService.Add(scheduleSpec, scheduleHandle, scheduleSlot);
            isTimerActive = true;
        }

        /// <summary>
        /// Stop observing.
        /// </summary>
        public virtual void StopObserve()
        {
            if (log.IsDebugEnabled)
            {
                log.Debug(".StopObserve");
            }

            if (isTimerActive)
            {
	            context.SchedulingService.Remove(scheduleHandle, scheduleSlot);
	            isTimerActive = false;
	            scheduleHandle = null;
            }
        }

        private static readonly Log log = LogFactory.GetLog(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
    }
}